test/501_migrations/normal_migrations.t

```diff
@@ -0,0 +1,53 @@
+The 501 parsetree contains a parsing modificacion.
+[compare_on.exe <file>] checks if there's a diff between the
+AST's resulting from
+1. parsing <file> on 5.1.0 directly
+2. parsing <file> on 5.1.0, migrating down to 5.0.0 and migrating back to 5.1.0
+We only expect a diff in one special case.
+
+  $ echo "let x : int = 5" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let _ : int = 5" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : type a b c. a -> b -> c = fun x y -> assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f = (fun (type a) (type b) (type c) -> (fun x y -> assert false : a -> b -> c))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let _ = (fun (type a) (type b) (type c) -> (fun x y -> assert false : a -> b -> c))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : type a . a -> a = fun x -> x" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let (x, y) : (int * int) = assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f: type a. a option -> _ = assert false" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+  $ echo "let f : 'a . 'a = (fun (type a) -> (assert false : a))" > file.ml
+  $ ./compare_on.exe file.ml ./identity_driver.exe
+
+Here we expect a diff (downwards migrating should yield the same as in the example right above).
+However, something is wrong.
```

Comment:
The tests are in the wrong direction: we expect that the round trip `r = 5.1 ⇒ 5.0 ⇒ 5.1` is a normalization.

Consequently, for every syntactic category `s` for which the down migration `down` is not onto, the reverse `up` migration should pick a best representative among the pre-images of `s` by `down` .

Since in our case, the best representative is the new syntactic construct for `let pat: constraint = exp`, we should expect a diff whenever we transform a non-normalized 5.1 syntactic form into a normalized form.

In other words, the diff should appear for
```
   $ echo "let ((x, y) : (int * int)) = assert false" > file.ml
   $ ./compare_on.exe file.ml ./identity_driver.exe


   $ echo 'let (x : [`A | `B]) = (`A :> [`A | `B])' > file.ml
   $ ./compare_on.exe file.ml ./identity_driver.exe

   $ echo 'let (x : [`A]) = (`A : [`A] :> [< `A | `B])'  > file.ml
   $ ./compare_on.exe file.ml ./identity_driver.exe
```
because the those syntactic forms and the new value constraint form map to the same 5.0 parsetree node, and the `up` transform will normalize then into the new syntactic form.

Reply:
Thanks a lot for the in-depth review, @Octachron, this is super helpful for us! And also thanks for all the explanations, among others the ones here.

I'm a bit confused about the statement

> The tests are in the wrong direction: we expect that the round trip `r = 5.1 ⇒ 5.0 ⇒ 5.1` is a normalization.

What do you mean with "The tests are in the wrong direction"? The tests in this file do test the round trip ` r = 5.1 ⇒ 5.0 ⇒ 5.1`. That's the round trip, for which I expect that there might be diffs. For the other round trip ` r2 = 5.0 ⇒ 5.1 ⇒ 5.0`, I don't expect diffs (up to the location problem I've mentioned) for the following reason: `down: 5.1 ⇒ 5.0` is onto, so we can choose `up: 5.0 ⇒ 5.1` to be its right inverse: `r2 = down \circ up = id`.

Btw, that's exactly the choice we've made: we've chosen `up` to be `down`'s right inverse. Unfortunately, that means that `r` isn't a normalization: the example you're mentioning `let ((x, y) : (int * int)) = assert false` is mapped to its syntactic form in `5.1` rather than its normal form. From a practical point of view that makes a lot of sense, I think: `r2` is by far the more important round trip than `r`: `r` will only happen during the compiler's `5.1` release phase. At the end of that phase, we'll bump the Ppxlib AST to `5.1` and from that moment on only `r2` can happen (and _will_ happen for everyone using ocmal <= 5.0 and a PPX). Do you think that was a bad choice of ours?

(As a side-note: I'm assuming there's a typo in

> Consequently, for every syntactic category `s` for which the down migration `down` is not onto

Did you want to say that `down` is _not_ onto or that it's onto? If you indeed want to say _not_ onto, could you give an example in `5.0` that isn't hit?)

Reply:
The two last commits on https://github.com/Octachron/ppxlib/tree/pr420_non_abstract_poly_fix fix the down and upward translation to avoid going through an invalid 5.0 parsetree. However, with the two commits we are still able to keen `r(⭨⭧)` the identity for
```ocaml
let f: 'a. 'a = (fun (type a) -> assert false:a)
```
because the test https://github.com/Octachron/ppxlib/blob/pr420_non_abstract_poly_fix/astlib/migrate_500_501.ml#L317 fails due to the difference in locations between the pattern and expression constraint.